### PR TITLE
redux, routing, data parsing

### PR DESCRIPTION
**Pull request by sndmax**

### CODE REVIEW

File: src/actions/post-actions.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import * as types from '../actions/action-types';
+
+export function getPostsSuccess(posts) {
```
Comment (f3arnil) [inline src/actions/post-actions.js]: Use ES6 syntax
const a = () => {}

File: src/components/layouts/main-layout.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { Link } from 'react-router-dom';
+
+export default function(props) {
```
Comment (f3arnil) [inline src/components/layouts/main-layout.js]: Lets follow one coomon style of components definition, ex src/components/home.js

File: src/components/views/post-list.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+export default (props) => {
```
Comment (f3arnil) [inline src/components/views/post-list.js]: Lets follow one coomon style of components definition, ex src/components/home.js

File: src/components/views/post-list.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    return (
+        <div>{props.posts.map(post => {
```
Comment (f3arnil) [inline src/components/views/post-list.js]: Define variables that will be used in the begining of render method 
ex. const { posts } = this.props
and then use just 'posts' variable

File: src/reducers/user-reducer.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import * as types from '../actions/action-types';
```
Comment (f3arnil) [inline src/reducers/user-reducer.js]: Do we need all types to be imported in user reducer?

File: src/components/containers/post-list-container.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import PostList from '../views/post-list';
+import * as postApi from '../../api/posts-api';
```
Comment (f3arnil) [inline src/components/containers/post-list-container.js]: Can we add aliases in webpack configuration to simplify work with pathes?

File: src/components/containers/post-list-container.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+class PostListContainer extends Component {
+
+    componentDidMount() {
+        postApi.getPosts();
+        //store.dispatch(getPostsSuccess('posts'))
```
Comment (f3arnil) [inline src/components/containers/post-list-container.js]: Remove commented code

File: src/components/containers/post-list-container.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    };
+
+    render() {
+        return (
+            <PostList posts={this.props.posts} />
```
Comment (f3arnil) [inline src/components/containers/post-list-container.js]: Define used variables in the begining of the method

File: src/actions/post-actions.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    return {
+        type: types.GET_POSTS_SUCCESS,
+        posts
```
Comment (f3arnil) [inline src/actions/post-actions.js]: Usually developers use payload property to provide data

File: src/components/layouts/main-layout.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import React, { Component } from 'react';
+import { Link } from 'react-router-dom';
+
+class MainLayout extends Component {
+    constructor(props) {
```
Comment (f3arnil) [inline src/components/layouts/main-layout.js]: I think there is no neccesety to describe constructor here

File: src/components/layouts/main-layout.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+            <div className="app">
+                <header>
+                    <Link className="app-logo" to="/"><img src="http://placehold.it/200x150" /></Link>
+                    <aside>
+                        <Link to="/about">About</Link>
+                        <Link to="/posts">Blog</Link>
+                    </aside>
+                </header>
+                <main id="page">
+                    {this.props.children}
```
Comment (f3arnil) [inline src/components/layouts/main-layout.js]: Define children variable in the begining of this metod

File: src/components/layouts/main-layout.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+                </main>
+            </div>
+        )
+    }
+}
+
+export default MainLayout;
```
Comment (f3arnil) [inline src/components/layouts/main-layout.js]: Add empty line

File: src/components/views/post-list.css [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  background-size: cover;
+  transform: scale(1.1);
+  filter: opacity(65%) blur(5px);
+  width: 100%;
+  height: 100%;
+  position: fixed;
+  left: 0;
+  right: 0;
+  z-index: -1;
+  background-image: url(https://static.pexels.com/photos/355770/pexels-photo-355770.jpeg); }
```
Comment (f3arnil) [inline src/components/views/post-list.css]: May be (in next PR) we should store images localy?

File: src/components/views/post-list.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import React, { Component } from 'react';
+import './post-list.scss';
+
+class PostList extends Component {
+    constructor(props) {
```
Comment (f3arnil) [inline src/components/views/post-list.js]: Do we need constructor here?

File: src/components/views/post-list.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+            <div>{posts.map(post => {
+                let date = new Date(post.date * 1000);
+
+                return (
+                    <article key={post.id} className="article-body">
+                        <header>
+                            <div className="article-date">{date.toDateString()}</div>
+                            <h1 className="article-title">{post.title}</h1>
+                            <div className="article-info">
+                                <a href="#">{post.author}</a> ● {post.tags.map(tag => {return <a href="#" key={tag.toString()}><span>{tag}</span></a>})}
```
Comment (f3arnil) [inline src/components/views/post-list.js]: I think that 'dot' symbole could be applied with styles ?
Make this line of code more readable moving its parts to the new lines

File: src/components/views/post-list.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+                            </div>
+                        </header>
+                        <div>
+                            <p>{post.content.substr(0, 300)}...</p>
+                            <p><a href="#" className="article-more">Read More →</a></p>
```
Comment (f3arnil) [inline src/components/views/post-list.js]: Arrow could be applied with css styles
Is # link correct?

File: src/components/views/post-list.scss [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+    color: $base-font-color;
+
+    &:hover {
+      color: $primary-font-color;
+    }
+  }
+}
+
+.article-title {
+  margin: 12px 0;
```
Comment (f3arnil) [inline src/components/views/post-list.scss]: For the future - we can implement mixin to generate margins and paddings

File: src/index.css [hunk 18]
```diff
@@ -1,7 +1,36 @@
...
+  background-size: cover;
+  transform: scale(1.1);
+  filter: grayscale(60%) blur(6px);
+  width: 100%;
+  height: 100%;
+  position: fixed;
+  left: 0;
+  right: 0;
+  z-index: -1;
+  background-image: url(https://static.pexels.com/photos/355770/pexels-photo-355770.jpeg); }
```
Comment (f3arnil) [inline src/index.css]: The samу as previous

**Comment by aeroslav** [comment]
*.map files should be removed from repo and added to .gitignore

File: src/components/layouts/main-layout.js [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    constructor(props) {
+        super(props);
+    }
+
+    render() {
+        return (
+            <div className="app">
+                <header>
+                    <Link className="app-logo" to="/"><img src="http://placehold.it/200x150" /></Link>
+                    <aside>
```
Comment (aeroslav) [inline src/components/layouts/main-layout.js]: Maybe you should create separate Nav component?

File: src/components/layouts/main-layout.js (hunk 12 above)
Comment (aeroslav) [reply src/components/layouts/main-layout.js]: There is a rule for ES Lint to check empty line at the EOF.

File: src/components/views/post-list.css [hunk 20]
```diff
@@ -0,0 +1,58 @@
+body {
```
Comment (aeroslav) [inline src/components/views/post-list.css]: Why there is a .scss and .css variants of this file? Can we remove .css one?

File: src/components/views/post-list.js [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    render() {
+        const { posts } = this.props;
+        return (
+            <div>{posts.map(post => {
+                let date = new Date(post.date * 1000);
+
+                return (
+                    <article key={post.id} className="article-body">
```
Comment (aeroslav) [inline src/components/views/post-list.js]: Extract Article as a separate component. There could be even 2 of them: for full article and for article preview.

File: src/components/views/post-list.js [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+
+class PostList extends Component {
+    constructor(props) {
+        super(props);
+    }
+
+    render() {
+        const { posts } = this.props;
+        return (
+            <div>{posts.map(post => {
```
Comment (aeroslav) [inline src/components/views/post-list.js]: `posts.map(({ id, date, title, author }) => {...})` will make your function look better

File: src/components/views/post-list.js (hunk 15 above)
Comment (aeroslav) [inline src/components/views/post-list.js]: Sometime, don't be afraid to extract even such small things as Tag into component, because later more features can appear for them. E.g. we can have static tag (just link) and editable tag, with 'x' icon on it.

File: src/components/views/post-list.js [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+                    <article key={post.id} className="article-body">
+                        <header>
+                            <div className="article-date">{date.toDateString()}</div>
+                            <h1 className="article-title">{post.title}</h1>
+                            <div className="article-info">
+                                <a href="#">{post.author}</a> ● {post.tags.map(tag => {return <a href="#" key={tag.toString()}><span>{tag}</span></a>})}
+                            </div>
+                        </header>
+                        <div>
+                            <p>{post.content.substr(0, 300)}...</p>
```
Comment (aeroslav) [inline src/components/views/post-list.js]: '300' is so-called magic number. Extract it into constant.

File: src/index.scss [hunk 24]
```diff
@@ -1,6 +1,48 @@
...
+$primary-spacing: 5px;
+
 body {
   margin: 0;
   padding: 0;
-  font-family: sans-serif; }
+  font-family: Merriweather;
+  font-weight: 300;
+  font-style: normal;
+  font-size: 15px;
```
Comment (aeroslav) [inline src/index.scss]: Use variables for font sizes and derive some font-based properties (like `line-height`) from this var.

File: src/index.scss [hunk 25]
```diff
@@ -1,6 +1,48 @@
...
+  background-size: cover;
+  transform: scale(1.1);
+  filter: opacity(65%) blur(5px);
+  width: 100%;
+  height: 100%;
+  position: fixed;
+  left: 0;
+  right: 0;
+  z-index: -1;
+  background-image: url(https://static.pexels.com/photos/355770/pexels-photo-355770.jpeg);
```
Comment (aeroslav) [inline src/index.scss]: Sort your properties by type, like here: https://css-tricks.com/poll-results-how-do-you-order-your-css-properties/

File: src/index.scss [hunk 26]
```diff
@@ -1,6 +1,48 @@
...
+#page {
+  -moz-box-sizing: border-box;
+  -webkit-box-sizing: border-box;
+  box-sizing: border-box;
+  width: auto;
+  margin: auto;
+  max-width: 1020px;
+  padding: 96px 32px;
+  -moz-osx-font-smoothing: auto;
+  -webkit-font-smoothing: subpixel-antialiased;
```
Comment (aeroslav) [inline src/index.scss]: Do we need prefixes in scss? Maybe add autoprefixer-loader?

File: src/reducers/post-reducer.js [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+    posts: [],
+    post: {
+        repos: []
+    }
+};
+
+const postReducer = (state = initialState, action) => {
+    switch (action.type) {
+        case GET_POSTS_SUCCESS:
+            return Object.assign({}, state, {posts: action.payload});
```
Comment (aeroslav) [inline src/reducers/post-reducer.js]: You can use destructuring `{ ...state, posts: action.payload }`

File: src/actions/post-actions.js [hunk 28]
```diff
@@ -0,0 +1,15 @@
+import { GET_POSTS_SUCCESS, POST_SUCCESS } from 'actions/action-types';
+
+export const getPostsSuccess = (posts) => {
+    return {
```
Comment (aeroslav) [inline src/actions/post-actions.js]: Use redux-thunk for this actions like this.

File: src/api/posts-api.js [hunk 29]
```diff
@@ -1,10 +1,32 @@
...
 import store from '../store';
-import { getPostsSuccess } from 'actions/post-actions';
+import { getPostsSuccess, getPostsWaiting, postSuccess, postWaiting } from 'actions/post-actions';
 
 export const getPosts = () => {
-    return axios.get('../posts.json', {responseType: 'json'})
+    return axios.get('../posts.json', { responseType: 'json' })
         .then(response => {
-            store.dispatch(getPostsSuccess(response.data.posts))
+            console.log('Waiting for data...');
```
Comment (f3arnil) [inline src/api/posts-api.js]: Remove console logs

File: src/api/posts-api.js [hunk 30]
```diff
@@ -1,10 +1,32 @@
...
+            store.dispatch(getPostsWaiting());
+
+            setTimeout(() => {
```
Comment (f3arnil) [inline src/api/posts-api.js]: Artificially delay the request response is not good solution.

File: src/api/posts-api.js [hunk 31]
```diff
@@ -1,10 +1,32 @@
...
+        });
+};
+
+export const getPost = (id) => {
+    return axios.get('../posts.json', { responseType: 'json' })
+        .then(response => {
+            console.log('Waiting for data...');
+            store.dispatch(postWaiting());
+
+            setTimeout(() => {
```
Comment (f3arnil) [inline src/api/posts-api.js]: The same

File: src/api/posts-api.js [hunk 32]
```diff
@@ -1,10 +1,32 @@
...
-import { getPostsSuccess } from 'actions/post-actions';
+import { getPostsSuccess, getPostsWaiting, postSuccess, postWaiting } from 'actions/post-actions';
 
 export const getPosts = () => {
-    return axios.get('../posts.json', {responseType: 'json'})
+    return axios.get('../posts.json', { responseType: 'json' })
         .then(response => {
-            store.dispatch(getPostsSuccess(response.data.posts))
+            console.log('Waiting for data...');
+            store.dispatch(getPostsWaiting());
```
Comment (f3arnil) [inline src/api/posts-api.js]: You should pass this method to components using connect, as a result you will not need to use store to dispatch action

File: src/components/containers/PostContainer.js [hunk 33]
```diff
@@ -0,0 +1,29 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import Article from 'views/Article';
+import PreLoader from 'views/PreLoader';
+import * as postApi from 'api/posts-api';
+
+class PostContainer extends Component {
+    componentDidMount() {
+        const { id } = this.props.match.params;
+        postApi.getPost(id);
```
Comment (f3arnil) [inline src/components/containers/PostContainer.js]: I think it will be better to use connect to provide this methods

File: src/components/views/Article/Article.js [hunk 34]
```diff
@@ -0,0 +1,29 @@
+import React, { Component } from 'react';
+import Tags from 'views/Tags';
+import './Article.scss';
+
+class Article extends Component {
+    render() {
+        const { id, date, title, author, tags, content } = this.props.post,
```
Comment (f3arnil) [inline src/components/views/Article/Article.js]: If there will be no post object in props you will get an error

File: src/components/views/Article/Article.js [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+            pub_date = new Date(date * 1000);
```
Comment (f3arnil) [inline src/components/views/Article/Article.js]: As I know defining variables with commas is bad practice. I suggest not to do it

File: src/components/views/Article/Article.js [hunk 36]
```diff
@@ -0,0 +1,29 @@
...
+
+        return (
+            <article className="article-body">
+                <header>
+                    <div className="article-date">{pub_date.toDateString()}</div>
```
Comment (f3arnil) [inline src/components/views/Article/Article.js]: 1. I think that for this case you can define variable fully at the line 8 or modify it out of return method - it will allow you to debug it easilly in future
2. You can think about using special libraries for dates manipulations

File: src/components/views/Articles/Articles.js [hunk 37]
```diff
@@ -0,0 +1,40 @@
...
+
+        return (
+            <section>
+            {posts.map(({ id, date, title, author, tags, content }) => {
+                let pub_date = new Date(date * 1000);
+
+                return (
+                    <article key={id.toString()} className="article-body">
+                        <header>
+                            <div className="article-date">{pub_date.toDateString()}</div>
```
Comment (f3arnil) [inline src/components/views/Articles/Articles.js]: The same as prev

File: src/components/views/Articles/Articles.js [hunk 38]
```diff
@@ -0,0 +1,40 @@
...
+                            <h1 className="article-title">{title}</h1>
+                            <div className="article-info">
+                                <a href="#">{author}</a>
+                                <span className="article-dot" />
+                                <Tags tags={tags} />
+                            </div>
+                        </header>
+                        <div>
+                            <p>{content.substr(0, symbolsAmount)}...</p>
```
Comment (f3arnil) [inline src/components/views/Articles/Articles.js]: The same as prev (definition out of return - its not critical, but I think it make sence)

File: src/components/views/PreLoader/PreLoader.js [hunk 39]
```diff
@@ -0,0 +1,53 @@
+import React, { Component } from 'react';
```
Comment (f3arnil) [inline src/components/views/PreLoader/PreLoader.js]: I think it is better to use external packages for loader (ex https://github.com/CognizantStudio/react-loader)  instead of implemented by yourself

File: src/components/views/PreLoader/index.js [hunk 40]
```diff
@@ -0,0 +1,3 @@
+import PreLoader from './PreLoader';
```
Comment (f3arnil) [inline src/components/views/PreLoader/index.js]: What?

File: src/reducers/user-reducer.js [hunk 41]
```diff
@@ -2,17 +2,15 @@ import { GET_USERS_SUCCESS, USER_PROFILE_SUCCESS } from 'actions/action-types';
...
 };
 
 const userReducer = (state = initialState, action) => {
     switch (action.type) {
         case GET_USERS_SUCCESS:
-            return Object.assign({}, state, {users: action.payload});
+            return { ...state, users: action.payload };
         case USER_PROFILE_SUCCESS:
-            return Object.assign({}, state, {userProfile: action.payload});
+            return { ...state, userProfile: action.payload };
```
Comment (f3arnil) [inline src/reducers/user-reducer.js]: Why did you delete Object.assign?

File: src/reducers/post-reducer.js [hunk 42]
```diff
@@ -1,18 +1,20 @@
...
 const postReducer = (state = initialState, action) => {
     switch (action.type) {
         case GET_POSTS_SUCCESS:
-            return Object.assign({}, state, {posts: action.payload});
+            return { ...state, posts: action.payload };
+        case GET_POSTS_WAITING:
+            return action.payload;
         case POST_SUCCESS:
-            return Object.assign({}, state, {post: action.payload});
+            return { ...state, post: action.payload };
```
Comment (f3arnil) [inline src/reducers/post-reducer.js]: Why did you delete Object.assign?

File: .gitignore [hunk 43]
```diff
@@ -22,3 +22,5 @@ yarn-error.log*
 
 .idea
 .eslintrc
+
+*.map
```
Comment (f3arnil) [inline .gitignore]: Add empty line in the end

File: src/components/views/Article/Article.js (hunk 34 above)
Comment (aeroslav) [reply src/components/views/Article/Article.js]: @f3arnil Well, Article is really need `post` prop, so I think it's ok to throw error if this component are used without providing it with necessary `post` data.

File: src/components/views/PreLoader/PreLoader.js (hunk 39 above)
Comment (aeroslav) [reply src/components/views/PreLoader/PreLoader.js]: @f3arnil Why it is better solution? I think it's not so crucial.

File: src/reducers/user-reducer.js (hunk 41 above)
Comment (aeroslav) [reply src/reducers/user-reducer.js]: Per my advice. Let's use ES2015+ 😉
